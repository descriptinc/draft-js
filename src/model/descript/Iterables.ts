/*
 * Iterables.ts
 * descript-web-v2
 *
 * Last modified by srubin on 3/26/19 7:12 PM
 * Copyright Â© 2019 Descript, Inc. All rights reserved.
 */

export function first<T>(iter: Iterable<T>): T | undefined {
  for (const t of iter) {
    return t;
  }
  return undefined;
}

export function last<T>(iter: Iterable<T>): T | undefined {
  let lastElt: T | undefined = undefined;
  for (const t of iter) {
    lastElt = t;
  }
  return lastElt;
}

export function size<T>(iter: Iterable<T>): number {
  let count = 0;
  for (const _ of iter) {
    count += 1;
  }
  return count;
}

export function filter<T, U extends T>(
  iter: Iterable<T>,
  predicate: (item: T, index: number) => item is U,
): IterableIterator<U>;
export function filter<T>(
  iter: Iterable<T>,
  predicate: (item: T, index: number) => boolean,
): IterableIterator<T>;
export function* filter<T>(
  iter: Iterable<T>,
  predicate: (item: T, index: number) => boolean,
): IterableIterator<T> {
  let index = 0;
  for (const item of iter) {
    if (predicate(item, index)) {
      yield item;
    }
    index += 1;
  }
}

export function* map<T, U>(
  iter: Iterable<T>,
  mapper: (item: T, index: number) => U,
): IterableIterator<U> {
  let i = 0;
  for (const item of iter) {
    yield mapper(item, i);
    i += 1;
  }
}

export function forEach<T>(
  iter: Iterable<T>,
  callback: (item: T, index: number) => void,
): void {
  let i = 0;
  for (const item of iter) {
    callback(item, i);
    i += 1;
  }
}

export function* flatMap<T, U>(
  iter: Iterable<T>,
  mapper: (item: T, index: number) => U | undefined,
): IterableIterator<U> {
  let index = 0;
  for (const item of iter) {
    const result = mapper(item, index);
    if (result !== undefined) {
      yield result;
    }
    index += 1;
  }
}

export function contains<T>(iter: Iterable<T>, item: T): boolean {
  for (const iterItem of iter) {
    if (item === iterItem) {
      return true;
    }
  }
  return false;
}

export function find<T>(
  iter: Iterable<T>,
  predicate: (item: T) => boolean,
): T | undefined {
  return first(filter(iter, predicate));
}

export function findIndex<T>(
  iter: Iterable<T>,
  predicate: (item: T) => boolean,
): number | undefined {
  let index = 0;
  for (const item of iter) {
    if (predicate(item)) {
      return index;
    }
    index += 1;
  }
  return undefined;
}

export function* slice<T>(
  iter: Iterable<T>,
  start: number,
  end: number = Infinity,
): IterableIterator<T> {
  let i = 0;
  for (const item of iter) {
    if (i >= end) {
      return;
    }
    if (i >= start) {
      yield item;
    }
    i += 1;
  }
}

export function some<T>(
  iter: Iterable<T>,
  predicate: (item: T) => boolean,
): boolean {
  return find(iter, predicate) !== undefined;
}
export function every<T>(
  iter: Iterable<T>,
  predicate: (item: T) => boolean,
): boolean {
  return !some(iter, item => !predicate(item));
}

export function isEmpty<T>(iter: Iterable<T>): boolean {
  return first(iter) === undefined;
}

export function takeNth<T>(iter: Iterable<T>, n: number): T | undefined {
  if (n <= 0) {
    throw new RangeError('n <= 0');
  }
  for (const item of iter) {
    n -= 1;
    if (n === 0) {
      return item;
    }
  }
  if (n > 1) {
    throw new RangeError(`less than n-1 items`);
  }
  return undefined;
}

export function* flatten<T>(iters: Iterable<Iterable<T>>): Iterable<T> {
  for (const iter of iters) {
    yield* iter;
  }
}

export function* withNext<T>(
  iter: Iterable<T>,
): IterableIterator<[T, T | undefined]> {
  let prev: T | undefined = undefined;
  for (const item of iter) {
    if (prev !== undefined) {
      yield [prev, item];
    }
    prev = item;
  }
  if (prev !== undefined) {
    yield [prev, undefined];
  }
}

export function* toIterator<T>(iter: Iterable<T>): IterableIterator<T> {
  yield* iter;
}

export function zip<T, U>(
  iterA: Iterable<T>,
  iterB: Iterable<U>,
): IterableIterator<[T, U]>;
export function zip<T, U, V>(
  iterA: Iterable<T>,
  iterB: Iterable<U>,
  iterC: Iterable<V>,
): IterableIterator<[T, U, V]>;
export function zip<T, U, V, W>(
  iterA: Iterable<T>,
  iterB: Iterable<U>,
  iterC: Iterable<V>,
  iterD: Iterable<W>,
): IterableIterator<[T, U, V, W]>;
export function* zip<T>(...iters: Iterable<T>[]): IterableIterator<T[]> {
  const iterators = iters.map(toIterator);
  while (true) {
    const res = [];
    for (const iterator of iterators) {
      const next = iterator.next();
      if (next.done) {
        return;
      }
      res.push(next.value);
    }
    yield res;
  }
}

export function* range(a: number, b?: number): IterableIterator<number> {
  let start = 0;
  let end = a;
  if (b !== undefined) {
    start = a;
    end = b;
  }
  for (let i = start; i < end; i++) {
    yield i;
  }
}

export function times<T>(
  count: number,
  callback: (i: number) => T,
): IterableIterator<T> {
  return map(range(count), callback);
}

export function join(iter: Iterable<string>, separator: string): string {
  let isFirst = true;
  let str = '';
  for (const item of iter) {
    if (isFirst) {
      str = item;
      isFirst = false;
    } else {
      str += separator + item;
    }
  }
  return str;
}

/**
 * Returns the equivalent of calling `Array.from(iterable)` but returns the
 * previous array if the elements are equal according to `isElementEqual`.
 *
 * Also preserves reference equality for matching elements
 * (i.e., previousArray[i] and incoming[i]) even if other
 * elements are changing.
 *
 * If you want to check reference equality of elements, do not pass
 * a value for `isElementEqual` - the function will be more efficient
 * than if you passed `(x, y) => x === y`
 */
export function toArrayOrPrevious<T>(
  incoming: Iterable<T>,
  previousArray: readonly T[],
  isElementEqual?: (a: T, b: T) => boolean,
): readonly T[] {
  if (incoming === previousArray) {
    return previousArray;
  }
  let res: T[] | undefined;
  let i = 0;
  const previousLength = previousArray.length;
  for (const item of incoming) {
    if (res) {
      let toPush = item;
      if (i < previousLength) {
        const prevItem = previousArray[i];
        if (item === prevItem || isElementEqual?.(item, prevItem)) {
          toPush = prevItem;
        }
      }
      res.push(toPush);
    } else if (
      i === previousLength ||
      !(isElementEqual
        ? isElementEqual(previousArray[i], item)
        : previousArray[i] === item)
    ) {
      if (Array.isArray(incoming) && !isElementEqual) {
        return incoming;
      }
      res = previousArray.slice(0, i);
      res.push(item);
    }
    i += 1;
  }
  if (!res) {
    if (i < previousArray.length) {
      if (Array.isArray(incoming) && !isElementEqual) {
        return incoming;
      }
      return previousArray.slice(0, i);
    }
    return previousArray;
  }
  return res;
}

/**
 * Returns a function that gets the latest result of the passed-in
 * iterable-returning function, using `toArrayOrPrevious` to memoize the result.
 */
export function makeMemoizedToArray<T, Arg0>(
  getIterable: (arg0: Arg0) => Iterable<T>,
  isElementEqual?: (a: T, b: T) => boolean,
): (arg0: Arg0) => readonly T[];
export function makeMemoizedToArray<T, Arg0, Arg1>(
  getIterable: (arg0: Arg0, arg1: Arg1) => Iterable<T>,
  isElementEqual?: (a: T, b: T) => boolean,
): (arg0: Arg0, arg1: Arg1) => readonly T[];
export function makeMemoizedToArray<T, Arg0, Arg1, Arg2>(
  getIterable: (arg0: Arg0, arg1: Arg1, arg2: Arg2) => Iterable<T>,
  isElementEqual?: (a: T, b: T) => boolean,
): (arg0: Arg0, arg1: Arg1, arg2: Arg2) => readonly T[];
export function makeMemoizedToArray<T, Arg0, Arg1, Arg2>(
  getIterable: (arg0: Arg0, arg1: Arg1, arg2: Arg2) => Iterable<T>,
  // defaults to reference equality
  isElementEqual?: (a: T, b: T) => boolean,
): (arg0: Arg0, arg1: Arg1, arg2: Arg2) => readonly T[] {
  let lastArray: readonly T[] | undefined;
  let lastArg0: Arg0 | undefined;
  let lastArg1: Arg1 | undefined;
  let lastArg2: Arg2 | undefined;
  return (arg0, arg1, arg2): readonly T[] => {
    // If inputs haven't changed, return the previous value
    if (
      lastArray &&
      arg0 === lastArg0 &&
      arg1 === lastArg1 &&
      arg2 === lastArg2
    ) {
      return lastArray;
    }
    const iter = getIterable(arg0, arg1, arg2);
    lastArg0 = arg0;
    lastArg1 = arg1;
    lastArg2 = arg2;
    if (!lastArray) {
      lastArray = Array.from(iter);
    } else {
      lastArray = toArrayOrPrevious(iter, lastArray, isElementEqual);
    }
    return lastArray;
  };
}
