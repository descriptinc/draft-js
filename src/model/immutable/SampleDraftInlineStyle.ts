/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

export const BOLD = new Set(['BOLD']);
export const BOLD_ITALIC = new Set(['BOLD', 'ITALIC']);
export const BOLD_ITALIC_UNDERLINE = new Set(['BOLD', 'ITALIC', 'UNDERLINE']);
export const BOLD_UNDERLINE = new Set(['BOLD', 'UNDERLINE']);
export const CODE = new Set(['CODE']);
export const ITALIC = new Set(['ITALIC']);
export const ITAIC_UNDERLINE = new Set(['ITALIC', 'UNDERLINE']);
export const NONE = new Set<string>();
export const STRIKETHROUGH = new Set(['STRIKETHROUGH']);
export const UNDERLINE = new Set(['UNDERLINE']);
