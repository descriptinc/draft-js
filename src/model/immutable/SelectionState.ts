/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

export type SelectionState = Readonly<{
  anchorKey: string;
  anchorOffset: number;
  focusKey: string;
  focusOffset: number;
  isBackward: boolean;
  hasFocus: boolean;
}>;

export function getStartOffset({
  isBackward,
  focusOffset,
  anchorOffset,
}: SelectionState): number {
  return isBackward ? focusOffset : anchorOffset;
}

export function getEndOffset({
  isBackward,
  anchorOffset,
  focusOffset,
}: SelectionState): number {
  return isBackward ? anchorOffset : focusOffset;
}

export function isCollapsed({
  anchorKey,
  focusKey,
  anchorOffset,
  focusOffset,
}: SelectionState): boolean {
  return anchorKey === focusKey && anchorOffset === focusOffset;
}

export function getStartKey({
  isBackward,
  focusKey,
  anchorKey,
}: SelectionState): string {
  return isBackward ? focusKey : anchorKey;
}

export function getEndKey({
  isBackward,
  focusKey,
  anchorKey,
}: SelectionState): string {
  return isBackward ? anchorKey : focusKey;
}

/**
 * Return whether the specified range overlaps with an edge of the
 * SelectionState.
 */
export function hasEdgeWithin(
  sel: SelectionState,
  blockKey: string,
  start: number,
  end: number,
): boolean {
  const {anchorKey, focusKey, anchorOffset, focusOffset} = sel;
  if (anchorKey === focusKey && anchorKey === blockKey) {
    const selectionStart = getStartOffset(sel);
    const selectionEnd = getEndOffset(sel);
    return (
      (start <= selectionStart && selectionStart <= end) || // selectionStart is between start and end, or
      (start <= selectionEnd && selectionEnd <= end) // selectionEnd is between start and end
    );
  }

  if (blockKey !== anchorKey && blockKey !== focusKey) {
    return false;
  }

  const offsetToCheck = blockKey === anchorKey ? anchorOffset : focusOffset;

  return start <= offsetToCheck && end >= offsetToCheck;
}

export function makeSelectionState({
  anchorKey,
  anchorOffset,
  focusKey,
  focusOffset,
  isBackward = false,
  hasFocus = false,
}: Partial<SelectionState> &
  Pick<
    SelectionState,
    'anchorKey' | 'anchorOffset' | 'focusKey' | 'focusOffset'
  >): SelectionState {
  return {
    anchorKey,
    anchorOffset,
    focusKey,
    focusOffset,
    isBackward,
    hasFocus,
  };
}

export function makeNullSelection(): SelectionState {
  return makeEmptySelection('');
}

export function makeEmptySelection(key: string): SelectionState {
  return makeSelectionState({
    anchorKey: key,
    anchorOffset: 0,
    focusKey: key,
    focusOffset: 0,
  });
}
