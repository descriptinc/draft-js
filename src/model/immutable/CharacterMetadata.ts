/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {DraftInlineStyle} from './DraftInlineStyle';
import {getKey2} from '../descript/cache';

export type CharacterMetadata = Readonly<{
  entity: string | null;
  style: DraftInlineStyle;
}>;

export function applyStyle(
  meta: CharacterMetadata,
  style: string,
): CharacterMetadata {
  if (meta.style.has(style)) {
    return meta;
  }
  return {
    ...meta,
    style: new Set([...meta.style.keys(), style]),
  };
}

export function removeStyle(
  meta: CharacterMetadata,
  style: string,
): CharacterMetadata {
  if (!meta.style.has(style)) {
    return meta;
  }
  const without = new Set(meta.style);
  without.delete(style);
  return {
    ...meta,
    style: without,
  };
}

export function applyEntity(
  meta: CharacterMetadata,
  entity: string | null,
): CharacterMetadata {
  if (meta.entity === entity) {
    return meta;
  }
  return {
    ...meta,
    entity,
  };
}

const pool: Record<string, CharacterMetadata> = {};
export const EMPTY_SET: DraftInlineStyle = new Set();
export const EMPTY_CHARACTER: CharacterMetadata = {
  entity: null,
  style: EMPTY_SET,
};

export function makeCharacterMetadata(
  config?: Partial<CharacterMetadata>,
): CharacterMetadata {
  if (!config) {
    return EMPTY_CHARACTER;
  }
  const entity = config.entity || null;
  const style = config.style || new Set();
  const styles = [...style];
  styles.sort(); // FIXME [perf]: is this sort too expensive?
  const styleStr = styles.join(',');
  const key = getKey2(entity || '', styleStr);
  const existing = pool[key];
  if (existing) {
    return existing;
  }
  const char = {entity, style};
  pool[key] = char;
  return char;
}
