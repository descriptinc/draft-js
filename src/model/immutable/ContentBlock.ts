/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {
  CharacterMetadata,
  EMPTY_CHARACTER,
  EMPTY_SET,
} from './CharacterMetadata';
import {BlockNode} from './BlockNode';
import {repeat} from '../descript/Iterables';
import {findRangesImmutable} from './findRangesImmutable';
import {DraftInlineStyle} from './DraftInlineStyle';
import fastDeepEqual from 'fast-deep-equal/es6';

const EMPTY_DATA = {};

export type ContentBlock = BlockNode;

export function makeContentBlock({
  data = EMPTY_DATA,
  key,
  text = '',
  depth = 0,
  type = 'unstyled',
  characterList = Array.from(repeat(text.length, EMPTY_CHARACTER)),
}: Partial<ContentBlock> & {key: string}): ContentBlock {
  return {
    characterList,
    data,
    key,
    text,
    depth,
    type,
  };
}

const haveEqualStyle = (
  charA: CharacterMetadata,
  charB: CharacterMetadata,
): boolean => fastDeepEqual(charA.style, charB.style);

const haveEqualEntity = (
  charA: CharacterMetadata,
  charB: CharacterMetadata,
): boolean => charA.entity === charB.entity;

export function findStyleRanges(
  block: BlockNode,
  filterFn: (value: CharacterMetadata) => boolean,
  callback: (start: number, end: number) => void,
): void {
  findRangesImmutable(block.characterList, haveEqualStyle, filterFn, callback);
}

export function findEntityRanges(
  block: BlockNode,
  filterFn: (value: CharacterMetadata) => boolean,
  callback: (start: number, end: number) => void,
): void {
  findRangesImmutable(block.characterList, haveEqualEntity, filterFn, callback);
}

export function getInlineStyleAt(
  block: BlockNode,
  index: number,
): DraftInlineStyle {
  return block.characterList[index]?.style || EMPTY_SET;
}

export function getEntityAt(block: BlockNode, index: number): string | null {
  return block.characterList[index]?.entity || null;
}
