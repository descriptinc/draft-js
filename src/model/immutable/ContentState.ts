/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {BlockMap} from './BlockMap';
import {
  SelectionState,
  makeEmptySelection,
  makeNullSelection,
} from './SelectionState';
import {first, last} from '../descript/Iterables';
import {BlockNodeRecord} from './BlockNodeRecord';
import {DraftEntityType} from '../entity/DraftEntityType';
import {DraftEntityMutability} from '../entity/DraftEntityMutability';
import {createFromArray} from './BlockMapBuilder';
import {sanitizeDraftText} from '../encoding/sanitizeDraftText';
import {makeContentBlockNode} from './ContentBlockNode';
import {makeContentBlock} from './ContentBlock';
import {gkx} from '../../stubs/gkx';
import DraftEntity from '../entity/DraftEntity';

export type ContentState = Readonly<{
  blockMap: BlockMap;
  selectionBefore: SelectionState;
  selectionAfter: SelectionState;
}>;

export function makeContentState({
  blockMap,
  selectionAfter,
  selectionBefore,
}: ContentState): ContentState {
  // FIXME [perf]: force type to be exact and don't create a new object?
  return {
    blockMap,
    selectionBefore,
    selectionAfter,
  };
}

export function getLastCreatedEntityKey(): string {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__getLastCreatedEntityKey();
}

export function getBlockForKey(
  {blockMap}: ContentState,
  key: string,
): BlockNodeRecord {
  const block = blockMap.get(key);
  if (!block) {
    throw new Error('No block found for key');
  }
  return block;
}

export function getFirstBlock({blockMap}: ContentState): BlockNodeRecord {
  const block = first(blockMap.values());
  if (!block) {
    throw new Error('Block map is empty');
  }
  return block;
}

export function getLastBlock({blockMap}: ContentState): BlockNodeRecord {
  const block = last(blockMap.values()); // FIXME [perf]: O(n)
  if (!block) {
    throw new Error('Block map is empty');
  }
  return block;
}

export function hasText(content: ContentState): boolean {
  const {blockMap} = content;
  if (blockMap.size > 1) {
    return true;
  }
  return (
    blockMap.size > 1 ||
    (blockMap.size > 0 &&
      escape(getFirstBlock(content).text).replace(/%u200B/g, '').length > 0)
  );
}

export function createEntity(
  type: DraftEntityType,
  mutability: DraftEntityMutability,
  data?: Object,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__create(type, mutability, data);
}

export function mergeEntityData(
  key: string,
  toMerge: Record<string, any>,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__mergeData(key, toMerge);
}

export function replaceEntityData(
  key: string,
  newData: Record<string, any>,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__replaceData(key, newData);
}

export function addEntity(instance: DraftEntityInstance): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__add(instance);
}

export function getEntity(key: string): DraftEntityInstance {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__get(key);
}

export function createFromBlockArray(
  blocks: BlockNodeRecord[] | {contentBlocks: BlockNodeRecord[]},
): ContentState {
  // TODO: remove this when we completely deprecate the old entity API
  const theBlocks = Array.isArray(blocks) ? blocks : blocks.contentBlocks;

  const blockMap = createFromArray(theBlocks);
  const firstBlock = first(blockMap.values());
  const selectionState = !firstBlock
    ? makeNullSelection()
    : makeEmptySelection(firstBlock.key);

  return makeContentState({
    blockMap,
    selectionBefore: selectionState,
    selectionAfter: selectionState,
  });
}

export function createFromText(
  text: string,
  delimiter: string | RegExp = /\r\n?|\n/g,
): ContentState {
  const strings = text.split(delimiter);
  const blocks = strings.map(block => {
    block = sanitizeDraftText(block);
    if (gkx('draft_tree_data_support')) {
      return makeContentBlockNode({
        text: block,
      });
    } else {
      return makeContentBlock({
        text: block,
      });
    }
  });
  return createFromBlockArray(blocks);
}
