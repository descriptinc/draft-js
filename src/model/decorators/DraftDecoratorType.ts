/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */

import {BlockNodeRecord} from '../immutable/BlockNodeRecord';
import {ContentState} from '../immutable/ContentState';

/**
 * An interface for document decorator classes, allowing the creation of
 * custom decorator classes.
 *
 * See `CompositeDraftDecorator` for the most common use case.
 */
export type DraftDecoratorType = {
  /**
   * Given a `ContentBlock`, return an immutable List of decorator keys.
   */
  getDecorations: (
    block: BlockNodeRecord,
    contentState: ContentState,
  ) => readonly (string | null)[];
  /**
   * Given a decorator key, return the component to use when rendering
   * this decorated range.
   */
  getComponentForKey: (key: string) => Function;
  /**
   * Given a decorator key, optionally return the props to use when rendering
   * this decorated range.
   */
  getPropsForKey: (key: string) => Object | null;
};
