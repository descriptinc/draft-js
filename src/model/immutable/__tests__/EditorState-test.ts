import {DraftDecoratorType} from '../../decorators/DraftDecoratorType';
import {makeSelectionState, SelectionState} from '../SelectionState';
import {makeContentBlock} from '../ContentBlock';
import {repeat} from '../../descript/Iterables';
import {makeCharacterMetadata} from '../CharacterMetadata';
import {BOLD, ITALIC} from '../SampleDraftInlineStyle';
import {
  acceptSelection,
  createEmpty,
  createWithContent,
  EditorState,
  getBlockTree,
  getCurrentInlineStyle,
  pushContent,
  setEditorState,
} from '../EditorState';
import {createFromBlockArray, getBlockForKey} from '../ContentState';
import RichTextEditorUtil from '../../modifier/RichTextEditorUtil';
import DraftModifier from '../../modifier/DraftModifier';

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

class Decorator implements DraftDecoratorType {
  getDecorations() {
    return [];
  }
  getComponentForKey(_: string) {
    return () => null;
  }
  getPropsForKey(_: string) {
    return null;
  }
}

Decorator.prototype.getDecorations = jest.fn();

const DEFAULT_SELECTION = {
  anchorKey: 'a',
  anchorOffset: 0,
  focusKey: 'a',
  focusOffset: 0,
  isBackward: false,
};

const collapsedSelection = makeSelectionState(DEFAULT_SELECTION);
const rangedSelection = makeSelectionState({
  ...DEFAULT_SELECTION,
  focusOffset: 1,
});

const plainBlock = makeContentBlock({
  key: 'a',
  text: 'Arsenal',
});
const boldBlock = makeContentBlock({
  key: 'b',
  text: 'Burnley',
  characterList: Array.from(repeat(7, makeCharacterMetadata({style: BOLD}))),
});
const boldA = Array.from(repeat(boldBlock.text.length, 'x'));
const emptyBlockA = makeContentBlock({
  key: 'emptyA',
  text: '',
});
const emptyBlockB = makeContentBlock({
  key: 'emptyB',
  text: '',
});
const italicBlock = makeContentBlock({
  key: 'c',
  text: 'Chelsea',
  characterList: Array.from(repeat(7, makeCharacterMetadata({style: ITALIC}))),
});

const getSampleEditorState = (type: string, decorator?: DraftDecoratorType) => {
  switch (type) {
    case 'DECORATED':
      return createWithContent(
        createFromBlockArray([boldBlock, italicBlock]),
        decorator,
      );
    case 'MULTI_BLOCK':
      return createWithContent(
        createFromBlockArray([emptyBlockA, emptyBlockB, boldBlock]),
      );
    case 'UNDECORATED':
    default:
      return createWithContent(
        createFromBlockArray([
          plainBlock,
          boldBlock,
          emptyBlockA,
          emptyBlockB,
          italicBlock,
        ]),
      );
  }
};

const UNDECORATED_STATE = getSampleEditorState('UNDECORATED');

const MULTI_BLOCK_STATE = getSampleEditorState('MULTI_BLOCK');

const assertGetCurrentInlineStyle = (
  selection: SelectionState,
  state: EditorState = UNDECORATED_STATE,
) => {
  const editorState = acceptSelection(state, selection);
  expect(getCurrentInlineStyle(editorState)).toMatchSnapshot();
};

beforeEach(() => {
  Decorator.prototype.getDecorations = jest.fn().mockImplementation(v => {
    return v === boldBlock
      ? boldA
      : Array.from(repeat(v.text.length, undefined));
  });
});

test('uses right of the caret at document start', () => {
  assertGetCurrentInlineStyle(collapsedSelection);
});

test('uses left of the caret, at position `1+`', () => {
  assertGetCurrentInlineStyle({
    ...collapsedSelection,
    anchorOffset: 1,
    focusOffset: 1,
  });
});

test('uses right of the caret at offset `0` within document', () => {
  assertGetCurrentInlineStyle({
    ...collapsedSelection,
    anchorKey: 'b',
    focusKey: 'b',
  });
});

test('uses previous block at offset `0` within empty block', () => {
  assertGetCurrentInlineStyle({
    ...collapsedSelection,
    anchorKey: 'emptyA',
    focusKey: 'emptyA',
  });
});

test('looks upward through empty blocks to find a character with collapsed selection', () => {
  assertGetCurrentInlineStyle({
    ...collapsedSelection,
    anchorKey: 'emptyB',
    focusKey: 'emptyB',
  });
});

test('does not discard style override when changing block type', () => {
  let editor = createEmpty();

  editor = RichTextEditorUtil.toggleInlineStyle(editor, 'BOLD');
  editor = RichTextEditorUtil.toggleBlockType(editor, 'test-block');

  expect(getCurrentInlineStyle(editor)).toMatchSnapshot();
});

test('does not discard style override when adjusting depth', () => {
  let editor = createEmpty();

  editor = RichTextEditorUtil.toggleInlineStyle(editor, 'BOLD');
  editor = RichTextEditorUtil.onTab({preventDefault: () => {}}, editor, 1);

  expect(getCurrentInlineStyle(editor)).toMatchSnapshot();
});

test('does not discard style override when splitting block', () => {
  let editor = createEmpty();

  editor = RichTextEditorUtil.toggleInlineStyle(editor, 'BOLD');
  editor = pushContent(
    editor,
    DraftModifier.splitBlock(editor.currentContent, editor.selection),
    'split-block',
  );

  expect(getCurrentInlineStyle(editor)).toMatchSnapshot();
});

test('uses right of the start for blocks with text', () => {
  assertGetCurrentInlineStyle({
    ...rangedSelection,
    focusKey: 'b',
  });
});

test('uses left of the start if starting at end of block', () => {
  const blockB = getBlockForKey(UNDECORATED_STATE.currentContent, 'b');
  assertGetCurrentInlineStyle({
    ...collapsedSelection,
    anchorKey: 'b',
    anchorOffset: blockB.text.length,
    focusKey: 'c',
    focusOffset: 3,
  });
});

test('looks upward through empty blocks to find a character', () => {
  assertGetCurrentInlineStyle({
    ...rangedSelection,
    anchorKey: 'emptyA',
    anchorOffset: 0,
    focusKey: 'c',
    focusOffset: 3,
  });
});

test('falls back to no style if in empty block at document start', () => {
  assertGetCurrentInlineStyle(
    makeSelectionState({
      anchorKey: 'emptyA',
      anchorOffset: 0,
      focusKey: 'c',
      focusOffset: 3,
      isBackward: false,
    }),
    MULTI_BLOCK_STATE,
  );
});

test('must set a new decorator', () => {
  const decorator = new Decorator();
  const editorState = getSampleEditorState('DECORATED', decorator);
  const boldB = Array.from(repeat(boldBlock.text.length, 'y'));

  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();

  Decorator.prototype.getDecorations.mockImplementation((v, c) => {
    return v === boldBlock
      ? boldB
      : Array.from(repeat(v.text.length, undefined));
  });

  class NextDecorator implements DraftDecoratorType {
    getDecorations() {
      return [];
    }
    getComponentForKey(_: string) {
      return () => null;
    }
    getPropsForKey(_: string) {
      return null;
    }
  }
  NextDecorator.prototype.getDecorations = jest.fn();

  const newDecorator = new NextDecorator();

  NextDecorator.prototype.getDecorations.mockImplementation((v, c) => {
    return v === boldBlock ? boldB : Array.from(repeat(v.text.length, 'a'));
  });

  const withNewDecorator = setEditorState(editorState, {
    decorator: newDecorator,
  });

  expect(withNewDecorator !== editorState).toMatchSnapshot();

  // Twice for the initial tree map generation, then twice more for
  // filter comparison.
  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();

  // Twice for filter comparison
  expect(newDecorator.getDecorations.mock.calls.length).toMatchSnapshot();

  expect(withNewDecorator.decorator === newDecorator).toMatchSnapshot();

  // Preserve block trees that had the same decorator list.
  expect(
    getBlockTree(editorState, boldBlock.key) ===
      getBlockTree(withNewDecorator, boldBlock.key),
  ).toMatchSnapshot();

  expect(
    getBlockTree(editorState, italicBlock.key) !==
      getBlockTree(withNewDecorator, italicBlock.key),
  ).toMatchSnapshot();
});

test('must call decorator with correct argument types and order', () => {
  const decorator = new Decorator();
  getSampleEditorState('DECORATED', decorator);
  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();
});

test('must correctly remove a decorator', () => {
  const decorator = new Decorator();
  const editorState = getSampleEditorState('DECORATED', decorator);
  const withNewDecorator = setEditorState(editorState, {decorator: null});

  expect(withNewDecorator !== editorState).toMatchSnapshot();
  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();
  expect(withNewDecorator.decorator).toMatchSnapshot();
});
