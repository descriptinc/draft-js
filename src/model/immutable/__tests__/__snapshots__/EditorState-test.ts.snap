// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not discard style override when adjusting depth 1`] = `
Set {
  "BOLD",
}
`;

exports[`does not discard style override when changing block type 1`] = `
Set {
  "BOLD",
}
`;

exports[`does not discard style override when splitting block 1`] = `
Set {
  "BOLD",
}
`;

exports[`falls back to no style if in empty block at document start 1`] = `Set {}`;

exports[`looks upward through empty blocks to find a character 1`] = `
Set {
  "BOLD",
}
`;

exports[`looks upward through empty blocks to find a character with collapsed selection 1`] = `
Set {
  "BOLD",
}
`;

exports[`must call decorator with correct argument types and order 1`] = `2`;

exports[`must correctly remove a decorator 1`] = `true`;

exports[`must correctly remove a decorator 2`] = `2`;

exports[`must correctly remove a decorator 3`] = `null`;

exports[`must set a new decorator 1`] = `2`;

exports[`must set a new decorator 2`] = `true`;

exports[`must set a new decorator 3`] = `4`;

exports[`must set a new decorator 4`] = `3`;

exports[`must set a new decorator 5`] = `true`;

exports[`must set a new decorator 6`] = `true`;

exports[`must set a new decorator 7`] = `true`;

exports[`uses left of the caret, at position \`1+\` 1`] = `Set {}`;

exports[`uses left of the start if starting at end of block 1`] = `
Set {
  "BOLD",
}
`;

exports[`uses previous block at offset \`0\` within empty block 1`] = `
Set {
  "BOLD",
}
`;

exports[`uses right of the caret at document start 1`] = `Set {}`;

exports[`uses right of the caret at offset \`0\` within document 1`] = `
Set {
  "BOLD",
}
`;

exports[`uses right of the start for blocks with text 1`] = `Set {}`;
